### PR TITLE
Works on the chat page

### PR DESCRIPTION
**Pull request by vlop11**
Works on adding the chat page which will be used for each group. For now, it is very basic and essentially functions as a forum page. As time goes on, I will

- require the user to be logged in
- show the user's display name along with their message
- filter messages by the group they belong to
- have the messages update every time a new message is posted (w/o refreshing)

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/ChatServlet.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Query query = new Query("Message").addSort("timestamp", SortDirection.ASCENDING);
+    PreparedQuery pq = datastore.prepare(query);
+
+    ImmutableList.Builder<String> builder = new ImmutableList.Builder<>();
+    for (Entity message : pq.asIterable()) {
+        builder.add((String) message.getProperty("message-text"));
+    }
```
Comment (courtneywong) [inline src/main/java/com/google/sps/servlets/ChatServlet.java]: Try using a Stream here: https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html

It'll probably look something like 
```ImmutableList<String> messages = pq.asIterable().stream().map(message -> (String) message.getProperty("message-text")).collect(toImmutableList());```

File: src/main/java/com/google/sps/servlets/ChatServlet.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+@WebServlet("/chat")
+public class ChatServlet extends HttpServlet{
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Query query = new Query("Message").addSort("timestamp", SortDirection.ASCENDING);
+    PreparedQuery pq = datastore.prepare(query);
```
Comment (courtneywong) [inline src/main/java/com/google/sps/servlets/ChatServlet.java]: Try to make variable names semantic. pq doesn't mean anything to me. Maybe messageQuery?

File: src/main/java/com/google/sps/servlets/ChatServlet.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    ImmutableList.Builder<String> builder = new ImmutableList.Builder<>();
+    for (Entity message : pq.asIterable()) {
+        builder.add((String) message.getProperty("message-text"));
+    }
+    ImmutableList<String> messages = builder.build();
+
+    // Data will be passed in as a list of messages in a map (needed for template)
+    Map<String, ImmutableList<String>> data = new HashMap<>();
+    data.put("Messages", messages);
```
Comment (courtneywong) [inline src/main/java/com/google/sps/servlets/ChatServlet.java]: Use an ImmutableMap here. A shortcut to create one would be `ImmutableMap.of("Messages", messages)`

File: src/main/java/com/google/sps/servlets/ChatServlet.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/chat")
+public class ChatServlet extends HttpServlet{
+
```
Comment (courtneywong) [inline src/main/java/com/google/sps/servlets/ChatServlet.java:28]: Can you pull the hardcoded constants such as "Message" "message-text", etc out into variables?

File: src/main/java/templates/chat.soy [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{namespace templates.chat}
+
+{template .chatPage}
+{@param Messages: list<string>}
```
Comment (courtneywong) [inline src/main/java/templates/chat.soy]: Param names should be lowercase.

**Review by courtneywong — CHANGES_REQUESTED** [review]

File: src/main/java/templates/chat.soy [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+            <form action="/chat" method="POST">
+                <input type="text" name="message-text">
+                <input type="submit">
+            </form>
+        </body>
+
+    </html>
+
+{/template}
```
Comment (ahlschlager) [inline src/main/java/templates/chat.soy]: new line

**Review by ahlschlager — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/ChatServlet.java (hunk 4 above)
Comment (courtneywong) [reply src/main/java/com/google/sps/servlets/ChatServlet.java:28]: I don't see the constants for this?

**Review by courtneywong — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/servlets/ChatServlet.java (hunk 4 above)
Comment (vlop11) [reply src/main/java/com/google/sps/servlets/ChatServlet.java:28]: ahh sorry when I changed the loop for a stream, I deleted the variable. Do you mean pull the message text into a String text = message.getProperty("message-text") like this?

Comment (courtneywong) [reply src/main/java/com/google/sps/servlets/ChatServlet.java:28]: The strings themselves should be constants. Like `private static final MESSAGE_TEXT_ATTRIBUTE = "message-text";`

File: src/main/java/com/google/sps/servlets/ChatServlet.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Calls query on all entities of type Message
+    Query messageQuery = new Query("Message").addSort("timestamp", SortDirection.ASCENDING);
+    PreparedQuery preparedMessageQuery = datastore.prepare(messageQuery);
+
+    // Creates list of the messages text
+    ImmutableList<String> messagesList = Streams.stream(preparedMessageQuery.asIterable()).map(message -> 
+        (String) message.getProperty("message-text")).collect(ImmutableList.toImmutableList());
```
Comment (courtneywong) [inline src/main/java/com/google/sps/servlets/ChatServlet.java]: Nit: you can import the static function separately. `import com.google.common.collect.ImmutableList.toImmutableList;`

**Review by courtneywong — CHANGES_REQUESTED** [review]

**Review by courtneywong — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/ChatServlet.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/chat")
+public class ChatServlet extends HttpServlet{
+    private static final String MESSAGE_TEXT_PROPERTY = "message-text";
+    private static final String TIMESTAMP_PROPERTY = "timestamp";
+    private static final String MESSAGE_KIND = "Message";
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (sporkert) [inline src/main/java/com/google/sps/servlets/ChatServlet.java:34]: Can you add a comment that explains high-level what these functions are doing? You can either add it on the doGet/doPost or just above the class name

**Review by sporkert — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/servlets/ChatServlet.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    // Creates list of the messages text
+    ImmutableList<String> messagesList = Streams.stream(preparedMessageQuery.asIterable()).map(message -> 
+        (String) message.getProperty(MESSAGE_TEXT_PROPERTY)).collect(toImmutableList());
+
+    // Data will be passed in as a list of messages in a map (needed for template)
+    ImmutableMap<String, ImmutableList<String>> data = ImmutableMap.of("messages", messagesList);
+
+    SoyFileSet sfs = SoyFileSet
+            .builder()
+            // File path starts in target/portfolio-1
```
Comment (sporkert) [inline src/main/java/com/google/sps/servlets/ChatServlet.java]: Don't leave comments in the middle of the chain. It's fine to put it above this chunk (aka above `SoyFileSet sfs = SoyFileSet`)

File: src/main/java/com/google/sps/servlets/ChatServlet.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/chat")
+public class ChatServlet extends HttpServlet{
+    private static final String MESSAGE_TEXT_PROPERTY = "message-text";
+    private static final String TIMESTAMP_PROPERTY = "timestamp";
+    private static final String MESSAGE_KIND = "Message";
+
+  /**
+   * When called, doGet will query all previously posted messages to the chat,
```
Comment (sporkert) [inline src/main/java/com/google/sps/servlets/ChatServlet.java:30]: nit: leave a space between the "*" and your comments, so `* When`

**Review by sporkert — COMMENTED** [review]
Left small comments, feel free to ping me once it's resolved and I'll go ahead and approve :)

**Review by sporkert — APPROVED** [review]

**Pull request merged**